src/Functions/errorCode.cpp

```diff
@@ -0,0 +1,87 @@
+#include <Functions/castTypeToEither.h>
+#include <Functions/IFunctionImpl.h>
+#include <Functions/FunctionFactory.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnString.h>
+#include <Core/Field.h>
+#include <unordered_map>
+
+#if !defined(ARCADIA_BUILD)
+#    include <Common/config_version.h>
+#endif
+
+extern std::unordered_map<int, const char *> auto_error_codes;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+/** errorCode() - returns the macro name for the error code.
+  */
+class FunctionErrorCode : public IFunction
+{
+public:
+    static constexpr auto name = "errorCode";
+    static FunctionPtr create(const Context &)
+    {
+        return std::make_shared<FunctionErrorCode>();
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & types) const override
+    {
+        if (!isNumber(types.at(0)))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "The argument of function {} must have simple numeric type, possibly Nullable", name);
+
+        return std::make_shared<DataTypeString>();
```

Comment:
Maybe consider returning LowCardinality(String)?

Reply:
Will it be useful for the function that return single string?
Looks like extra conversion when there absolutely no point int his (if it will be added as materialized alias, then it should be wrapped with LowCardinality() - agree)

Actually I though about generating Enum for it, but this will complicate things, and I'm not sure that it worth it.

Reply:
Yes. It is applicable not only for constant expressions, for example, you can process query_log with this function.
Let's do it if it's easy to do.

Reply:
Ok, let's make it LowCardinality, but FWIW plain LowCardinality type has 2x overhead (but this is absolutely synthetic test):

- w/o low cardinality

```sql
SELECT toTypeName(errorCodeToName(0)) 
   
┌─toTypeName(errorCodeToName(0))─┐    
│ String                         │    
└────────────────────────────────┘  

SELECT errorCodeToName(number % 3000)                               
FROM system.numbers                             
LIMIT 1000000000.                                                   
FORMAT Null              

0 rows in set. Elapsed: 13.464 sec. Processed 1.00 billion rows, 8.00 GB (74.27 million rows/s., 594.20 MB/s.)    

SELECT errorCodeToName(number % 3000) 
FROM system.numbers
LIMIT 1000000000.
FORMAT Null
   
0 rows in set. Elapsed: 26.579 sec. Processed 1.00 billion rows, 8.00 GB (37.63 million rows/s., 301.01 MB/s.) 
```

- w/ LowCardinality:

```sql
SELECT toTypeName(errorCodeToName(0))

┌─toTypeName(errorCodeToName(0))─┐
│ LowCardinality(String)         │
└────────────────────────────────┘

SELECT errorCodeToName(number % 3000)
FROM system.numbers
LIMIT 1000000000.
FORMAT Null

0 rows in set. Elapsed: 26.579 sec. Processed 1.00 billion rows, 8.00 GB (37.63 million rows/s., 301.01 MB/s.) 
```

Here some profile numbers:

- w/o lowcardinality:

```
Samples: 69K of event 'cycles', 4000 Hz, Event count (approx.): 8859192325 lost: 0/0 drop: 0/0
Overhead  Shared Object        Symbol
  30.16%  clickhouse.fasttest  [.] DB::FunctionErrorCodeToName::executeImpl                                                                                                                                                                                                      ▒
  16.93%  clickhouse.fasttest  [.] DB::(anonymous namespace)::ModuloByConstantImpl<unsigned long, unsigned short>::vectorConstant                                                                                                                                                ▒
  15.82%  clickhouse.fasttest  [.] errorCodeToName                                                                                                                                                                                                                               ◆
  14.83%  clickhouse.fasttest  [.] DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::resize<>                                                                                                                                                                  ▒
   9.97%  libc-2.32.so         [.] __memmove_avx_unaligned_erms                                                                                                                                                                                                                  ▒
   3.24%  clickhouse.fasttest  [.] DB::ColumnVector<unsigned short>::getInt                                                                                                                                                                                                      ▒
   2.34%  clickhouse.fasttest  [.] DB::(anonymous namespace)::NumbersSource::generate
```

- w/o lowcardinality:

```
Samples: 97K of event 'cycles', 4000 Hz, Event count (approx.): 21324285943 lost: 0/0 drop: 0/0                                                                                                                                                                                   
Overhead  Shared Object           Symbol                                                                                                                                                                                                                                          
  19.80%  clickhouse.fasttest.lc  [.] DB::ColumnUnique<DB::ColumnString>::uniqueInsertRangeImpl<unsigned short>                                                                                                                                                                  ◆
  16.43%  clickhouse.fasttest.lc  [.] DB::ColumnString::insertData                                                                                                                                                                                                               ▒
  12.16%  clickhouse.fasttest.lc  [.] DB::ReverseIndex<unsigned long, DB::ColumnString>::insert                                                                                                                                                                                  ▒
   7.94%  clickhouse.fasttest.lc  [.] DB::FunctionErrorCodeToName::executeImpl                                                                                                                                                                                                   ▒
   7.91%  clickhouse.fasttest.lc  [.] DB::(anonymous namespace)::ModuloByConstantImpl<unsigned long, unsigned short>::vectorConstant                                                                                                                                             ▒
   7.39%  clickhouse.fasttest.lc  [.] DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::resize<>                                                                                                                                                               ▒
   6.62%  clickhouse.fasttest.lc  [.] DB::ErrorCodes::getName                                                                                                                                                                                                                    ▒
   5.18%  libc-2.32.so            [.] __strlen_avx2                                                                                                                                                                                                                              ▒
   4.20%  libc-2.32.so            [.] __memmove_avx_unaligned_erms                                                                                                                                                                                                               ▒
   3.61%  clickhouse.fasttest.lc  [.] memequalSSE2Wide                                                                                                                                                                                                                           ▒
   1.60%  clickhouse.fasttest.lc  [.] DB::ColumnVector<unsigned short>::getInt                                                                                                                                                                                                   ▒
   1.27%  clickhouse.fasttest.lc  [.] DB::(anonymous namespace)::NumbersSource::generate                                                                                                                                                                                         ▒

```

(Also this version has strlen, it wasn't optimized by the compiler, will address this bit)

P.S.
- benchmarks are not that accurate (use a little bit different implementations, but the profiler shows the idea) and purely synthetic
- and yes LowCardinality type can have benefits later, but still.

Reply:
>(Also this version has strlen, it wasn't optimized by the compiler, will address this bit)

0cd79de632047a3cbecd829616abb1456ca3a96e